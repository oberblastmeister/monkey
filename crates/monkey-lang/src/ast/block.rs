use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub open_bracket: T!['{'],
    pub stmts: Vec<ast::Stmt>,
    pub close_bracket: T!['}'],
}

impl Parse for Block {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        let open_bracket = p.parse()?;

        let mut stmts = Vec::new();

        loop {
            if p.is_eof()? || p.peek::<BlockEnd>() {
                break;
            }

            stmts.push(p.parse()?);
        }

        let close_bracket = p.parse()?;

        Ok(Block {
            open_bracket,
            stmts,
            close_bracket,
        })
    }
}

struct BlockEnd;

impl Peek for BlockEnd {
    fn peek(p: &mut Peeker) -> bool {
        matches!(p.nth(0), K!['}'])
    }
}
