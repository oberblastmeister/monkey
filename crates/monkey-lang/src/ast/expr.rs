use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Binary(Box<ast::ExprBinary>),
    Lit(ast::Lit),
}

impl Parse for Expr {
    #[inline]
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        expr(p)
    }
}

#[inline]
fn expr(p: &mut Parser) -> ParseResult<Expr> {
    expr_bp(p, 0)
}

fn expr_bp(p: &mut Parser, min_bp: u8) -> ParseResult<Expr> {
    let mut lhs = match p.nth(0)? {
        K!['('] => expr_paren(p)?,
        _ if p.peek::<ast::Lit>() => Expr::Lit(p.parse()?),
        t => panic!("Bad token: {:?}", t),
    };

    loop {
        if p.peek::<ExprEnd>() {
            break;
        };

        let op = ast::BinOp::from_peeker(&mut p.peeker).ok_or_else(|| {
            ParseError::expected(&p.next().unwrap(), "Expected a binary operator")
        })?;

        let (l_bp, r_bp) = op.binding_power();

        if l_bp < min_bp {
            break;
        }

        lhs = {
            let token = p.next()?;

            let rhs = expr_bp(p, r_bp)?;

            Expr::Binary(Box::new(ast::ExprBinary {
                lhs,
                op,
                rhs,
                token,
            }))
        };

        continue;
    }

    Ok(lhs)
}

struct ExprEnd;

impl Peek for ExprEnd {
    fn peek(p: &mut Peeker) -> bool {
        matches!(p.nth(0), K![eof] | K![;] | K![')'])
    }
}

fn expr_paren(p: &mut Parser) -> ParseResult<Expr> {
    assert_eq!(p.nth(0).unwrap(), K!['(']);

    p.next().unwrap();
    let lhs = expr_bp(p, 0)?;
    p.parse::<T![')']>()?;
    Ok(lhs)
}
