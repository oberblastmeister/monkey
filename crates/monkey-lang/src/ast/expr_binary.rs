use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprBinary {
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    pub op: BinOp,
    pub rhs: ast::Expr,
}

impl ExprBinary {
    pub fn with_lhs(
        p: &mut Parser<'_>,
        lhs: ast::Expr,
        op: BinOp,
    ) -> ParseResult<ExprBinary> {
        assert!(BinOp::from_peeker(&mut p.peeker).is_some());

        let token = p.next()?;

        Ok(ExprBinary {
            lhs,
            token,
            op,
            rhs: p.parse()?,
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
}

use BinOp::*;

impl BinOp {
    pub fn from_peeker(p: &mut Peeker<'_>) -> Option<BinOp> {
        Some(match p.nth(0) {
            K![+] => Add,
            K![-] => Sub,
            K![*] => Mul,
            K![/] => Div,
            _ => return None,
        })
    }

    pub fn binding_power(self) -> (u8, u8) {
        match self {
            Add | Sub => (1, 2),
            Mul | Div => (3, 4),
        }
    }
}

impl Peek for BinOp {
    fn peek(p: &mut Peeker) -> bool {
        BinOp::from_peeker(p).is_some()
    }
}
