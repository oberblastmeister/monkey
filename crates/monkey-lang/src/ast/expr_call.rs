use crate::{ast, Parse, ParseResult, Parser};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprCall {
    ident: T![ident],
    arguments: Arguments,
}

impl Parse for ExprCall {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        Ok(ExprCall {
            ident: p.parse()?,
            arguments: p.parse()?,
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Arguments(ast::List<T![ident], T!['{'], T!['}']>);

impl Parse for Arguments {
    fn parse(p: &mut Parser) -> ParseResult<Self> {
        Ok(Arguments(p.parse()?))
    }
}
