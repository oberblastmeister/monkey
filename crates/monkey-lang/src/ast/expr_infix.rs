use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprInfix {
    /// The left hand side of the expression
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
    /// The right hand side of the expression
    pub rhs: ast::Expr,
}

expr_parse!(Infix, ExprInfix, "binary expression");
