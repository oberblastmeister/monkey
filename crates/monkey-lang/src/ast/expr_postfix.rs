use crate::{ast, Parse, ParseError, ParseResult, Parser, Peek, Peeker, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub struct ExprPostfix {
    /// The right hand side of the expression
    pub lhs: ast::Expr,
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
}

expr_parse!(Postfix, ExprPostfix, "postfix expression");
