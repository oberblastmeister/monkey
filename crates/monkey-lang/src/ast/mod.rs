#[macro_use]
pub mod generated;

pub use generated::TokenKind;

macro_rules! expr_parse {
    ($ty:ident, $local:ty, $expected:literal) => {
        impl crate::Parse for $local {
            fn parse(p: &mut crate::Parser<'_>) -> Result<Self, crate::ParseError> {
                let span = p.tok_at(0)?.span;

                match crate::ast::Expr::parse(p)? {
                    crate::ast::Expr::$ty(expr) => Ok(*expr),
                    _ => Err(crate::ParseError::msg(&span, $expected)),
                }
            }
        }
    };
}

mod token;

mod expr;
mod expr_lit;
mod expr_infix;
mod expr_prefix;
mod expr_postfix;
mod expr_call;

mod lit;
mod lit_bool;
mod lit_num;
mod lit_str;
mod lit_fn;

mod stmt;
mod stmt_expr;
mod stmt_return;
mod stmt_let;
mod stmt_if;

mod block;

mod file;

mod op;

mod list;

pub use token::Token;

pub use expr::Expr;
pub use expr_lit::ExprLit;
pub use expr_infix::ExprInfix;
pub use expr_prefix::ExprPrefix;
pub use expr_postfix::ExprPostfix;
pub use expr_call::ExprCall;

pub use op::Op;

pub use lit::Lit;
pub use lit_bool::LitBool;
pub use lit_num::LitNum;
pub use lit_str::LitStr;
pub use lit_fn::LitFn;

pub use stmt::Stmt;
pub use stmt_expr::StmtExpr;
pub use stmt_return::StmtReturn;
pub use stmt_let::StmtLet;
pub use stmt_if::StmtIf;

pub use block::Block;

pub use file::File;

pub use list::List;
