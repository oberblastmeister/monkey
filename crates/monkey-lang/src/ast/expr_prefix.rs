use crate::ast;

#[derive(Debug, Clone, PartialEq)]
pub struct ExprPrefix {
    /// the token of the operator
    pub token: ast::Token,
    /// The operator type
    pub op: ast::Op,
    /// The right hand side of the expression
    pub rhs: ast::Expr,
}

expr_parse!(Prefix, ExprPrefix, "prefix expression");
