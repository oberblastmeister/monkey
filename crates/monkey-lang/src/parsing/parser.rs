use std::collections::VecDeque;

use super::{Lexer, Parse, ParseError, ParseErrorKind, ParseResult, Peek};
use crate::ast::{self, Token, TokenKind};
use crate::Span;

/// Construct used to peek a parser.
#[derive(Debug)]
pub struct Peeker<'a> {
    pub(crate) lexer: Lexer<'a>,

    buf: VecDeque<Token>,

    last: Option<Span>,
}

impl<'a> Peeker<'a> {
    pub fn new(input: &'a str) -> Peeker<'a> {
        let lexer = Lexer::new(input);
        let buf = VecDeque::new();
        let last = None;
        Peeker { lexer, buf, last }
    }

    /// Peek the token kind at the given position.
    pub fn nth(&mut self, n: usize) -> TokenKind {
        match self.at(n) {
            Ok(t) => match t {
                Some(t) => t.kind,
                None => TokenKind::Eof,
            },
            Err(error) => TokenKind::Error,
        }
    }

    /// Make sure there are at least `n` items in the buffer, and return the
    /// item at that point.
    fn at(&mut self, n: usize) -> ParseResult<Option<&Token>> {
        while self.buf.len() <= n {
            let token = match self.lexer.next_token() {
                Ok(token) => token,
                Err(e) if e.kind() == &ParseErrorKind::Eof => break,
                Err(e) => return Err(e),
            };

            self.last = Some(token.span);
            self.buf.push_back(token);
        }

        Ok(self.buf.get(n))
    }

    /// Peek something else
    pub fn peek<T>(&mut self) -> bool
    where
        T: Peek,
    {
        T::peek(self)
    }
}

#[derive(Debug)]
pub struct Parser<'a> {
    pub peeker: Peeker<'a>,
    errors: Vec<ParseError>,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        Parser {
            peeker: Peeker::new(input),
            errors: Vec::new(),
        }
    }

    /// Parse a specific item from the parser.
    pub fn parse<T>(&mut self) -> ParseResult<T>
    where
        T: Parse,
    {
        T::parse(self)
    }

    pub fn peek<T>(&mut self) -> bool
    where
        T: Peek,
    {
        T::peek(&mut self.peeker)
    }

    /// Try to consume a single thing matching `T`, returns `true` if any tokens
    /// were consumed.
    pub fn try_parse<T>(&mut self) -> ParseResult<bool>
    where
        T: Parse + Peek,
    {
        Ok(if self.peek::<T>() {
            self.parse::<T>()?;
            true
        } else {
            false
        })
    }

    /// Try to consume all things matching `T`, returns `true` if any tokens
    /// were consumed.
    pub fn try_parse_all<T>(&mut self) -> ParseResult<bool>
    where
        T: Parse + Peek,
    {
        let mut consumed = false;

        while self.peek::<T>() {
            self.parse::<T>()?;
            consumed = true;
        }

        Ok(consumed)
    }

    /// Consume the next token from the parser.
    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> ParseResult<Token> {
        if let Some(t) = self.peeker.buf.pop_front() {
            return Ok(t);
        }

        match self.peeker.lexer.next_token() {
            Err(e) if e.kind() == &ParseErrorKind::Eof => Ok(Token {
                span: Span::new(0, 0),
                kind: TokenKind::Eof,
                text: "".into(),
            }),
            Err(e) => Err(e),
            Ok(t) => Ok(t),
        }
    }

    /// Peek the token kind at the given position.
    pub fn nth(&mut self, n: usize) -> ParseResult<TokenKind> {
        if let Some(t) = self.peeker.at(n)? {
            Ok(t.kind)
        } else {
            Ok(TokenKind::Eof)
        }
    }

    /// Test if the parser is at end-of-file, after which there is no more input
    /// to parse.
    pub fn is_eof(&mut self) -> Result<bool, ParseError> {
        Ok(self.peeker.at(0)?.is_none())
    }

    /// Assert that the parser has reached its end-of-file.
    pub fn eof(&mut self) -> Result<(), ParseError> {
        if let Some(token) = self.peeker.at(0)? {
            return Err(ParseError::new(
                token,
                ParseErrorKind::ExpectedEof {
                    actual: token.kind.as_str(),
                },
            ));
        }

        Ok(())
    }
}

pub fn parse<T: Parse>(input: &str) -> ParseResult<T> {
    Parser::new(input).parse::<T>()
}

#[cfg(test)]
mod tests {
    use std::fs;

    use super::*;

    #[test]
    fn parser() {
        insta::glob!("snapshot_inputs/parser/*.txt", |path| {
            let input = fs::read_to_string(path).unwrap();
            let suffix = path.file_stem().unwrap().to_str().unwrap();
            insta::with_settings!({snapshot_path => "snapshots/parser", snapshot_suffix => suffix}, {
                insta::assert_debug_snapshot!(parse::<ast::File>(&input).map(|f| f.stmts))
            })
        })
    }
}
