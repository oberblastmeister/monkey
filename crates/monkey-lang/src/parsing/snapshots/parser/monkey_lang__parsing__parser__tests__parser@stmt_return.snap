---
source: crates/monkey-lang/src/parsing/parser.rs
expression: "parse::<ast::File>(&input).map(|f| f.stmts)"

---
Ok(
    [
        Return(
            StmtReturn {
                return_token: ReturnKw {
                    token: Token {
                        span: Span {
                            start: 0,
                            end: 6,
                        },
                        kind: ReturnKw,
                        text: "return",
                    },
                },
                expr: Lit(
                    Num(
                        LitNum {
                            num: Number {
                                token: Token {
                                    span: Span {
                                        start: 7,
                                        end: 11,
                                    },
                                    kind: Number,
                                    text: "1243",
                                },
                            },
                            value: 1243.0,
                        },
                    ),
                ),
                semi: Semicolon {
                    token: Token {
                        span: Span {
                            start: 11,
                            end: 12,
                        },
                        kind: Semicolon,
                        text: ";",
                    },
                },
            },
        ),
        Return(
            StmtReturn {
                return_token: ReturnKw {
                    token: Token {
                        span: Span {
                            start: 13,
                            end: 19,
                        },
                        kind: ReturnKw,
                        text: "return",
                    },
                },
                expr: Lit(
                    Str(
                        LitStr {
                            token: Token {
                                span: Span {
                                    start: 20,
                                    end: 35,
                                },
                                kind: String,
                                text: "\"return string\"",
                            },
                        },
                    ),
                ),
                semi: Semicolon {
                    token: Token {
                        span: Span {
                            start: 35,
                            end: 36,
                        },
                        kind: Semicolon,
                        text: ";",
                    },
                },
            },
        ),
    ],
)
